File: SyncRNG_modules.R
# Using SyncRNG to replicate 
library('SyncRNG')
library(MASS)
library('mvtnorm')
library(ggplot2)

universal_seed <- 123
universal_rng <- SyncRNG(seed=universal_seed)

syncrng.box.muller <- function(mu, sigma, n, seed=0, rng=NULL) {
  "Generate random samples from a normal distribution using the Box-Muller transform.
  
  Args:
     mu: The mean of the normal distribution.
     sigma: The standard deviation of the normal distribution.
     n: The number of random samples to generate.
     seed: An optional seed value for the random number generator (default is 0).
     rng: An optional instance of SyncRNG. If NULL, a new instance will be created using the seed.
  
   Returns:
     A vector of n random samples from the specified normal distribution."

  rng <- if (is.null(rng)) universal_rng else rng
  
  two.pi <- 2 * pi
  ngen <- ceiling(n / 2)
  out <- replicate(2 * ngen, 0.0)
  
  for (i in 1:ngen) {
    u1 <- 0.0
    u2 <- 0.0
    
    while (u1 == 0) { u1 <- rng$rand(); }
    while (u2 == 0) { u2 <- rng$rand(); }
    
    mag <- sigma * sqrt(-2.0 * log(u1))
    z0 <- mag * cos(two.pi * u2) + mu
    z1 <- mag * sin(two.pi * u2) + mu
    
    out[2*i - 1] = z0;
    out[2*i] = z1;
  }
  return(out[1:n]);
}


generate_random_matrix <- function(size, mu, sigma, rng=NULL) {
  "Generate a matrix with elements sampled from a normal distribution.

  Args:
    size: The dimension of the square matrix (number of rows and columns).
    mu: The mean of the normal distribution.
    sigma: The standard deviation of the normal distribution.
    seed: A seed value for the random number generator.

  Returns:
    A size x size matrix with elements sampled from the specified normal distribution."
  
  rng <- if (is.null(rng)) universal_rng else rng
  random_numbers <- syncrng.box.muller(mu, sigma, size * size, rng = rng)
  matrix(random_numbers, nrow = size, ncol = size)
}


generate_mvn_samples <- function(mean_vector, cov_matrix, rng=NULL) {
  "Generate a random vector from a multivariate normal distribution.

  Args:
    size: The dimension of the multivariate normal distribution (length of the mean vector).
    mean_vector: The mean vector of the multivariate normal distribution.
    cov_matrix: The covariance matrix of the multivariate normal distribution.
    seed: A seed value for the random number generator to ensure reproducibility.

  Returns:
    A random vector sampled from the specified multivariate normal distribution."
  
  # Cholesky decomposition of the covariance matrix
  rng <- if (is.null(rng)) universal_rng else rng
  size <- length(mean_vector)
  L <- t(chol(cov_matrix))  # Returns upper triangular matrix by default. Transpose to match Python. 
  Z <- syncrng.box.muller(0, 1, size, rng = rng)
  mvn_samples <- as.vector(L %*% Z + mean_vector)
  return(mvn_samples)
}

# mean_vector <- c(1, 1)
# cov_matrix <- diag(c(2, 2))  # Zero on off-diagonals --> no correlation between x and y
# print(generate_mvn_samples(mean_vector=mean_vector, cov_matrix=cov_matrix))
# 
# print(generate_random_matrix(2, 0, 1))

################################## EXAMPLES ####################################
###################### CHECK IF OUTPUT MATCHES PYTHON ##########################

# mu <- 0
# sigma <- 1
# size <- 3
# mean_vector <- c(1, 2, 3)
# cov_matrix <- matrix(c(1, 0.5, 0.3,
#                        0.5, 1, 0.4,
#                        0.3, 0.4, 1), nrow=3, ncol=3)
# 
# # Generate random numbers
# random_samples <- syncrng.box.muller(mu, sigma, 10)
# print("Random Samples:")
# print(random_samples)
# 
# # Generate a random matrix
# random_matrix <- generate_random_matrix(size, mu, sigma)
# print("Random Matrix:")
# print(random_matrix)
# 
# # Generate multivariate normal samples
# mvn_sample <- generate_mvn_samples(mean_vector, cov_matrix)
# print("Multivariate Normal Sample:")
# print(mvn_sample)

#################### COMPARE R WITH SYNCRNG FOR BASIC PLOTS ####################
# n_samples <- 1000
# s <- SyncRNG(seed=42)
# sync_uniforms <- replicate(n_samples, s$rand())
# base_uniforms <- runif(n_samples)
# 
# par(mfrow = c(2, 1))
# hist(sync_uniforms, breaks = 30, main = "SyncRNG Uniform Distribution", xlab = "Value")
# hist(base_uniforms, breaks = 30, main = "Base R Uniform Distribution", xlab = "Value")
# 
# # Compare Normal Random Numbers
# sync_normals <- syncrng.box.muller(0, 1, n_samples, seed = 123)
# base_normals <- rnorm(n_samples)
# 
# par(mfrow = c(2, 1))
# hist(sync_normals, breaks = 30, main = "SyncRNG Normal Distribution", xlab = "Value")
# hist(base_normals, breaks = 30, main = "Base R Normal Distribution", xlab = "Value")


#################### COMPARE MVN SAMPLES FROM SYNCRNG AND RMVNORM ##############
# Plot rmvnorm vs syncrng samples to compare multivariate normal distribution
# mean_vector <- c(1, 1)
# cov_matrix <- matrix(c(1, 0.5, 0.5, 1), nrow = 2)  # Zero on off-diagonals --> no correlation between x and y
# num_samples <- 1000
# 
# samples_rmvnorm <- rmvnorm(num_samples, mean_vector, cov_matrix)
# samples_syncrng <- matrix(0, nrow = num_samples, ncol = size)
# 
# for (i in 1:num_samples) {
#   samples_syncrng[i, ] <- generate_mvn_samples(mean_vector, cov_matrix)
# }
# 
# samples_rmvnorm_df <- data.frame(x = samples_rmvnorm[,1], y = samples_rmvnorm[,2])
# samples_syncrng_df <- data.frame(x = samples_syncrng[,1], y = samples_syncrng[,2])
# 
# samples_rmvnorm_df$source <- "rmvnorm"
# samples_syncrng_df$source <- "syncrng"
# combined_df <- rbind(samples_rmvnorm_df, samples_syncrng_df)
# 
# ggplot(combined_df, aes(x = x, y = y, color = source)) +
#   geom_point(alpha = 0.5) +
#   labs(title = "Comparison of Multivariate Normal Distribution Samples",
#        x = "X-axis",
#        y = "Y-axis") +
#   theme_minimal() +
#   scale_color_manual(values = c("rmvnorm" = "blue", "syncrng" = "red"))


#################### PLOT MVN SAMPLES FROM ONLY SYNCRNG ########################
# mean_vector <- c(0, 0)
# cov_matrix <- matrix(c(1, 0.9, 0.9, 1), nrow = 2)
# n_samples <- 100
# samples <- matrix(0, nrow = 2, ncol = n_samples)
# 
# for (i in 1:n_samples) {
#   samples[, i] <- generate_mvn_samples(mean_vector, cov_matrix)
# }
# 
# plot(samples[1, ], samples[2, ], main = "Bivariate normal with variance 1, covariance 0.9", asp = 1)

File: SyncRNG_modules.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Jul 25 02:03:26 2024

@author: mac
"""

import numpy as np
import math
import matplotlib.pyplot as plt
from SyncRNG import SyncRNG

# seed -> for single use
# rng -> for 'global' use, consistent rand. num. gen across multiple function calls
# set a universal seed or rng? will have to add 'rng' as parameter to all functions

universal_seed = 123
universal_rng = SyncRNG(seed=universal_seed)

def syncrng_box_muller(mu, sigma, n, seed=0, rng=None):
    """Generate n numbers from N(mu, sigma^2)"""
    rng = universal_rng if rng is None else rng

    two_pi = 2 * math.pi
    ngen = math.ceil(n / 2)
    out = [0.0] * 2 * ngen

    for i in range(ngen):
        u1 = 0.0
        u2 = 0.0

        while u1 == 0:
            u1 = rng.rand()
        while u2 == 0:
            u2 = rng.rand()

        mag = sigma * math.sqrt(-2.0 * math.log(u1))
        z0 = mag * math.cos(two_pi * u2) + mu
        z1 = mag * math.sin(two_pi * u2) + mu

        out[2*i] = z0
        out[2*i + 1] = z1

    return out[:n]

#plt.hist(syncrng_box_muller(mu=0, sigma=1, n=100), bins=10)

def generate_random_matrix(size, mu, sigma, rng=None):
    rng = universal_rng if rng is None else rng
    random_numbers = syncrng_box_muller(mu, sigma, size * size, rng=rng)
    return np.array(random_numbers).reshape(size, size)

def generate_mvn_samples(mean_vector, cov_matrix, rng=None):
    rng = universal_rng if rng is None else rng
    size = len(mean_vector)
    L = np.linalg.cholesky(cov_matrix).T
    Z = syncrng_box_muller(0, 1, size, rng=rng)
    mvn_samples = np.dot(L, Z) + mean_vector
    return mvn_samples

############################ EXAMPLES ############################
mean_vector = np.array([1, 3])
cov_matrix = np.array([[2, 0], [0, 2]])
n_samples = 1000
n_iterations = 3

mvn_sample = generate_mvn_samples(mean_vector, cov_matrix, rng=None)
print(mvn_sample)

random_sample = generate_random_matrix(2, 0, 1)
print(random_sample)

'''
# PLOT 1
for i in range(n_iterations):
    # Generate samples using generate_mvn_samples
    samples_custom = np.array([generate_mvn_samples(mean_vector, cov_matrix) for _ in range(n_samples)])

    # Generate samples using np.random.multivariate_normal
    samples_np = np.random.multivariate_normal(mean_vector, cov_matrix, n_samples)
    
    # Plot
    fig, ax = plt.subplots(1, 3, figsize=(18, 6))
    
    # Plot syncrng samples
    ax[0].scatter(samples_custom[:, 0], samples_custom[:, 1], alpha=0.5, label='Custom MVN', color='blue')
    ax[0].set_title(f'Custom MVN Samples - Iteration {i+1}')
    ax[0].set_xlabel('X')
    ax[0].set_ylabel('Y')
    ax[0].legend()
    
    # Plot numpy samples
    ax[1].scatter(samples_np[:, 0], samples_np[:, 1], alpha=0.5, label='NumPy MVN', color='orange')
    ax[1].set_title(f'NumPy MVN Samples - Iteration {i+1}')
    ax[1].set_xlabel('X')
    ax[1].set_ylabel('Y')
    ax[1].legend()
    
    # Plot both overlapped
    ax[2].scatter(samples_custom[:, 0], samples_custom[:, 1], alpha=0.5, label='Custom MVN', color='blue')
    ax[2].scatter(samples_np[:, 0], samples_np[:, 1], alpha=0.5, label='NumPy MVN', color='orange')
    ax[2].set_title(f'Overlapped MVN Samples - Iteration {i+1}')
    ax[2].set_xlabel('X')
    ax[2].set_ylabel('Y')
    ax[2].legend()
    
    plt.show()

'''
'''
# PLOT 2
fig, axes = plt.subplots(n_iterations, 3, figsize=(18, 6 * n_iterations))

for i in range(n_iterations):
    # Generate samples using generate_mvn_samples
    samples_custom = np.array([generate_mvn_samples(mean_vector, cov_matrix) for _ in range(n_samples)])

    # Generate samples using np.random.multivariate_normal
    samples_np = np.random.multivariate_normal(mean_vector, cov_matrix, n_samples)
    
    # Plot syncrng samples
    ax_custom = axes[i, 0]
    ax_custom.scatter(samples_custom[:, 0], samples_custom[:, 1], alpha=0.5, label='Custom MVN', color='blue')
    ax_custom.set_title(f'Custom MVN Samples - Iteration {i+1}')
    ax_custom.set_xlabel('X')
    ax_custom.set_ylabel('Y')
    ax_custom.legend()
    
    # Plot numpy samples
    ax_np = axes[i, 1]
    ax_np.scatter(samples_np[:, 0], samples_np[:, 1], alpha=0.5, label='NumPy MVN', color='orange')
    ax_np.set_title(f'NumPy MVN Samples - Iteration {i+1}')
    ax_np.set_xlabel('X')
    ax_np.set_ylabel('Y')
    ax_np.legend()
    
    # Plot both overlapped
    ax_overlap = axes[i, 2]
    ax_overlap.scatter(samples_custom[:, 0], samples_custom[:, 1], alpha=0.5, label='Custom MVN', color='blue')
    ax_overlap.scatter(samples_np[:, 0], samples_np[:, 1], alpha=0.5, label='NumPy MVN', color='orange')
    ax_overlap.set_title(f'Overlapped MVN Samples - Iteration {i+1}')
    ax_overlap.set_xlabel('X')
    ax_overlap.set_ylabel('Y')
    ax_overlap.legend()

plt.tight_layout()
plt.show()
'''
